cloudfoundry/uptimer
# Tcp availability

Adding a test for uptime of tcp routse.

This change deploys an app that listens for tcp traffic, and the measurement client makes a tcp connection to the app and looks for a specific string response ("Hello from Uptimer.")

The tas-runtime team is adding this so that we can use Uptimer in CI when deploying changes to routing release. 

---

To test:
1. Build uptimer; `go install` in the `uptimer` directory
1. Procure and target a cf-d environment
1. Save the value of `credhub get -n /bosh-[ENV]/cf/cf_admin_password` in the config below under the key` "admin_password"`
1. Run `bosh manifest > manifest.yml` and make a trivial change to the manifest, for example increasing an instance count of the api instance group.
1. Save the following config as `uptimer-config.json`:
```
 {
     "while": [
         {
             "command": "bosh",
             "command_args": ["deploy", "-d", "cf", "manifest.yml, "-n"]
         }
     ],
     "cf": {
         "api": "https://api.[ENV].cf-app.com",
         "app_domain": "[ENV].cf-app.com",
         "admin_user": "admin",
         "admin_password": "[PASSWORD]",
         "tcp_domain": "tcp.[ENV].cf-app.com",
         "use_single_app_instance": false,
         "available_port": 1025,
         "tcp_port": 1026
     },
     "optional_tests": {
       "run_tcp_availability": true,
       "run_app_syslog_availability": true
     },
     "allowed_failures": {
         "app_pushability": 2,
         "http_availability": 5,
         "tcp_availability": 5,
         "recent_logs": 2,
         "streaming_logs": 2,
         "app_syslog_availability": 2
     }
 }
```
1. Run `uptimer -configFile uptimer-config.json`

Expected result: Uptimer should do a bosh deploy of the saved manifest. You should see output that includes TCP measurements:

```
[UPTIMER] 2023/01/10 01:05:47 SUCCESS (HTTP availability): 0 failed attempts to perform get requests did not exceed the threshold of 5 allowed failures (Total attempts: 252, pass rate 100.00%)

[UPTIMER] 2023/01/10 01:05:47 Received stop signal for measurement App syslog availability

[UPTIMER] 2023/01/10 01:05:47 Received stop signal for measurement TCP availability

[UPTIMER] 2023/01/10 01:05:47 SUCCESS (App pushability): 0 failed attempts to push and delete an app did not exceed the threshold of 2 allowed failures (Total attempts: 3, pass rate 100.00%)

[UPTIMER] 2023/01/10 01:05:47 SUCCESS (Recent logs): 0 failed attempts to fetch recent logs did not exceed the threshold of 2 allowed failures (Total attempts: 26, pass rate 100.00%)

[UPTIMER] 2023/01/10 01:05:47 SUCCESS (Streaming logs): 2 failed attempts to stream logs did not exceed the threshold of 2 allowed failures (Total attempts: 8, pass rate 75.00%)

[UPTIMER] 2023/01/10 01:05:47 SUCCESS (TCP availability): 0 failed attempts to perform netcat requests did not exceed the threshold of 5 allowed failures (Total attempts: 252, pass rate 100.00%)

[UPTIMER] 2023/01/10 01:05:47 SUCCESS (App syslog availability): 0 failed attempts to check application syslogs did not exceed the threshold of 2 allowed failures (Total attempts: 8, pass rate 100.00%)
```
